karapace/schema_models.py

```diff
@@ -180,6 +183,50 @@ def schema(self) -> Draft7Validator | AvroSchema | ProtobufSchema:
         return parsed_typed_schema.schema
 
 
+class AvroMerge:
+    def __init__(self, schema_str: str, dependencies: Mapping[str, Dependency] | None = None):
+        self.schema_str = json_encode(json_decode(schema_str), compact=True, sort_keys=True)
+        self.dependencies = dependencies
+        self.unique_id = 0
+
+    def union_safe_schema_str(self, schema_str: str) -> str:
+        # in case we meet union - we use it as is
+        regex = re.compile(r"^\s*\[")
+        if regex.match(schema_str):
+            return (
+                "{"
+                + f' "name": "___un__ique_x_n_q_karapace____{self.unique_id}",'
+                + '"type":"record","fields":[{"name":"name", "type":'
+                + schema_str
+                + "}]}"
+            )
+
+        return (
+            "{"
+            + f' "name": "___un__ique_x_n_q_karapace____{self.unique_id}",'
+            + '"type":"record","fields":[{"name":"name", "type": '
+            + "["
+            + schema_str
+            + "]}]}"
+        )
+
+    def builder(self, schema_str: str, dependencies: Mapping[str, Dependency] | None = None) -> str:
+        """To support references in AVRO we recursively merge all referenced schemas with current schema"""
+        if dependencies:
+            merged_schema = ""
+            for dependency in dependencies.values():
+                merged_schema += self.builder(dependency.schema.schema_str, dependency.schema.dependencies) + ",\n"
```

Comment:
Can we please change this to a loop based approach supported by a stack or a queue, (depending on the propagation). This is to safe guard against a large nested schema.

Reply:
done